feat(i18n, canvas, context, side-panel): Duplication of tables

nikelborm opened the pull request:
## Gives ability to duplicate table using simple button

| From canvas | From side-panel |
|---|---|
| ![Screenshot From 2024-11-10 01-37-35](https://github.com/user-attachments/assets/47e899a9-784a-4734-84cd-b6f95d0cbed7) | ![Screenshot From 2024-11-10 01-40-11](https://github.com/user-attachments/assets/b962c3d5-df0c-4579-a598-9b5c10c7f917) |

## New table will have the same reconstructed indexes and columns, as original table, but with new IDs and creationDates for everything: the table, it's columns, and it's indexes, which point to according IDs of freshly created columns.

|Source table showing first index | Source table showing second index|
|---|---|
| ![Screenshot From 2024-11-10 01-48-13](https://github.com/user-attachments/assets/bd166cbf-839a-40c7-8758-228446c5a5fc) | ![Screenshot From 2024-11-10 01-48-02](https://github.com/user-attachments/assets/69ab215c-bc86-4053-9d17-075683398d52) |

|Duplicated table showing first index | Duplicated table showing second index|
|---|---|
| ![Screenshot From 2024-11-10 01-48-40](https://github.com/user-attachments/assets/a8677d8b-d10b-489b-bfcd-914fc4db13f3) | ![Screenshot From 2024-11-10 01-48-49](https://github.com/user-attachments/assets/bdcc7538-1462-4e2b-af1b-c9e33d733244) |

## Pressing those buttons leads to creation of a table's copy nearby

![Screenshot From 2024-11-10 01-37-44](https://github.com/user-attachments/assets/96be29e1-92fe-416a-8d4f-d16657cd5504)

## Relationships with other tables are not preserved

![Screenshot From 2024-11-10 02-01-09](https://github.com/user-attachments/assets/e22f7fd9-b600-49b9-bf1b-6e75f3b6ac74)

![Screenshot From 2024-11-10 02-01-15](https://github.com/user-attachments/assets/4693db31-5a59-4eb1-aedf-07534cfc41b1)

nikelborm commented:
Also I refactored i18n a bit in a way, that the only place you'll have to register new files with translations is in 
[/src/i18n/locales/index.ts](https://github.com/nikelborm/chartdb/blob/feature/duplicate-tables/src/i18n/locales/index.ts). And translation file will be automatically distributed in according `i18n.init` call parameters.

nikelborm commented:
Also painted in the same red color button to delete table that's shown after clicking on according node in canvas.

nikelborm commented:
Also refactored components a bit, and left comments why the change was made.

nikelborm commented:
Seems like a good idea to add:
1. Duplication of outcoming references, like foreign keys
2. Do something with index names, because if set manually, and duplicated with table, will probably clash, because use the same name
3. Mentioned in TODOs typescript checks, to ensure we don't accidentally share an array or object reference between duplicated table, because it may cause a problem such as when we change one table, they both change. Currently there're no such cases, as far as typescript tells us, but will be good idea, to add a safety net

nikelborm commented:
Generally this PR is ready to be review as it is, is doesn't break anything

nikelborm commented:
Rebased to latest upstream/main

guyb1 commented:
Hey @nikelborm, thanks so much for your hard work on this!

Since there’s already logic for duplicating tables and diagrams through the template clone flow, I think it would be best to reuse that existing implementation here. I’ve created this PR you can see: #404.

Thanks again for your help!

nikelborm commented:
I honestly don't know how to react. I was initially going to respond with how I feel about this situation using only [I-messages](https://en.m.wikipedia.org/wiki/I-message). I tried to do it with all my might, but there are things in this situation that are impossible to discuss only this way. And I stayed as constructive as I could.

I'm very upset after this.

Not only because this PR was here for five days without any feedback in a way that was psychologically blocking me from starting working on consequent PRs. I came every day to my PRs rebasing changes to make them as seamless to merge and review as possible. I updated and refined the description of PR to be as easy to understand as possible.

Not only because part of the refactoring that was done just along the way but wasn't strongly connected to the feature itself was just completely dismissed in a blink of an eye. And hadn't I moved part of the refactoring to a different PR you'll most likely dismiss it along the way too.

I'm also upset because you didn't ask me "Hey here is this and that, please use the new function I created a day ago". I still would have a chance to improve my PR and apply given feedback even if I still may not be too happy about this, I will actually just add the changes you asked and will compromise. And when my fixed PR will be merged I would feel like I have done something good and improved experience of the whole community and would be encouraged to do this again. But after this message I don't have this feeling.

The most upsetting part is that you took the work without any attribution by making commits from your name. I consider this stealing.

I initially had very big plans about my contributions to the project. I already wrote a list of things that need improvement and which I was planning to improve. Here are few examples from my notes:

> TODO: make t (for translation) typesafe in app.chartdb.io
> 
> TODO: slowly make app.chartdb.io more typesafe by enabling stricter rules in tsconfig and upgrading it's version
> 
> TODO: finish 3 little tasks in app.chartdb.io https://github.com/chartdb/chartdb/pull/378#issuecomment-2466540931
> 
> TODO: implement all combinations of more finegrained cardinality on both sides in chartdb:
> - 0/1/M
> - 0/1
> - 0/M
> - 1/M
> - 0
> - 1
> - M
> 
> And write SQL demos of what all of them will be converted to.
> 
> Also to not make it very confusing, there's also some standard like strictly 1 on the first side (1 and only 1)
> 
> TODO: change that stupid switch in a setting, so it automatically changes the menu in chartdb
> 
> I don't remember exactly what the name of the option is on the menu.
> 
> TODO some highlighting AFAIR when working with columns in index setting works with them by name, and not by the id of a field
> app.chartdb.io
> 
> TODO problem highlighter menu, which colors in red color clashes of table names, index names, and field names
> app.chartdb.io
> 
> TODO: remove security vulnerability from https://github.com/chartdb/chartdb/blob/main/Dockerfile that uses ARG
> 
> TODO: add AMOLED theme to chartdb
> 
> TODO: chartdb add an ability to share a diagram using a URL to GitHub gist. and interface for users to login locally right in the browser and inline choose with whom will this diagram be shared. No backend required
> 
> TODO: chartdb remove snap-to-grid-feature highlighting when holding shift and inside active focused text field
> 
> TODO: find an fix the reason why chartdb doesn't work as expected when using deleting with backspace. It doesn't print to the console that removeTable was called. I'm trying to understand how it works there, where the point where it reacts to pressing backspace
> 
> TODO: chartdb возможность вывести срендеренный текстовый список какие таблицы могут быть созданы без каких-либо значений в них — типа
> вы уверены, что вы хотите создать таблицу ресурс instance без abstract resource
> 
> TODO: better comments support chartdb. Allow to export them using postgres COMMENT ON syntax
> 
> TODO: add to chartdb template with my educational platform
> 
> TODO: add to chartdb template with my smart home
> 
> TODO: add to chartdb template with my trading firm
> 
> TODO: add to chartdb template with my fish system

Now I feel discouraged. I have the right to feel and express my emotions, and not everything here is purely my emotions. I hope you'll see the feedback lying in it too. I believe that every other new contributor that will get into such situation will feel the same. I don't believe that's the way you build community.

johnnyfish commented:
> I honestly don't know how to react. I was initially going to respond with how I feel about this situation using only [I-messages](https://en.m.wikipedia.org/wiki/I-message). I tried to do it with all my might, but there are things in this situation that are impossible to discuss only this way. And I stayed as constructive as I could.
> 
> I'm very upset after this.
> 
> Not only because this PR was here for five days without any feedback in a way that was psychologically blocking me from starting working on consequent PRs. I came every day to my PRs rebasing changes to make them as seamless to merge and review as possible. I updated and refined the description of PR to be as easy to understand as possible.
> 
> Not only because part of the refactoring that was done just along the way but wasn't strongly connected to the feature itself was just completely dismissed in a blink of an eye. And hadn't I moved part of the refactoring to a different PR you'll most likely dismiss it along the way too.
> 
> I'm also upset because you didn't ask me "Hey here is this and that, please use the new function I created a day ago". I still would have a chance to improve my PR and apply given feedback even if I still may not be too happy about this, I will actually just add the changes you asked and will compromise. And when my fixed PR will be merged I would feel like I have done something good and improved experience of the whole community and would be encouraged to do this again. But after this message I don't have this feeling.
> 
> The most upsetting part is that you took the work without any attribution by making commits from your name. I consider this stealing. I hadn't been mentioned in either commit description as a form of @mention or using special "coauthored-with" git comment, or PR description
> 
> I initially had very big plans about my contributions to the project. I already wrote a list of things that need improvement and which I was planning to improve. Here are few examples from my notes:
> 
> > TODO: make t (for translation) typesafe in app.chartdb.io
> > TODO: slowly make app.chartdb.io more typesafe by enabling stricter rules in tsconfig and upgrading it's version
> > TODO: finish 3 little tasks in app.chartdb.io [#378 (comment)](https://github.com/chartdb/chartdb/pull/378#issuecomment-2466540931)
> > TODO: implement all combinations of more finegrained cardinality on both sides in chartdb:
> > 
> > * 0/1/M
> > * 0/1
> > * 0/M
> > * 1/M
> > * 0
> > * 1
> > * M
> > 
> > And write SQL demos of what all of them will be converted to.
> > Also to not make it very confusing, there's also some standard like strictly 1 on the first side (1 and only 1)
> > TODO: change that stupid switch in a setting, so it automatically changes the menu in chartdb
> > I don't remember exactly what the name of the option is on the menu.
> > TODO some highlighting AFAIR when working with columns in index setting works with them by name, and not by the id of a field
> > app.chartdb.io
> > TODO problem highlighter menu, which colors in red color clashes of table names, index names, and field names
> > app.chartdb.io
> > TODO: remove security vulnerability from https://github.com/chartdb/chartdb/blob/main/Dockerfile that uses ARG
> > TODO: add AMOLED theme to chartdb
> > TODO: chartdb add an ability to share a diagram using a URL to GitHub gist. and interface for users to login locally right in the browser and inline choose with whom will this diagram be shared. No backend required
> > TODO: chartdb remove snap-to-grid-feature highlighting when holding shift and inside active focused text field
> > TODO: find an fix the reason why chartdb doesn't work as expected when using deleting with backspace. It doesn't print to the console that removeTable was called. I'm trying to understand how it works there, where the point where it reacts to pressing backspace
> > TODO: chartdb возможность вывести срендеренный текстовый список какие таблицы могут быть созданы без каких-либо значений в них — типа
> > вы уверены, что вы хотите создать таблицу ресурс instance без abstract resource
> > TODO: better comments support chartdb. Allow to export them using postgres COMMENT ON syntax
> > TODO: add to chartdb template with my educational platform
> > TODO: add to chartdb template with my smart home
> > TODO: add to chartdb template with my trading firm
> > TODO: add to chartdb template with my fish system
> 
> Now I feel discouraged. I have the right to feel and express my emotions, and not everything here is purely my emotions. I hope you'll see the feedback lying in it too. I believe that every other new contributor that will get into such situation will feel the same. I don't believe that's the way you build community.

@nikelborm, Thank you so much for your thoughtful message and for the incredible work you've done on your recent PR. We genuinely appreciate your contributions and the feature request you brought forward - it’s a fantastic improvement for the UX of ChartDB. Your dedication to refining and improving the PR is evident, and it hasn’t gone unnoticed.

We’ve had an intense few days of work on the product, and I recognize that some of the feedback and interactions, especially from @guyb1, may not have come across as constructively as intended. Our main goal was to ensure consistency in the codebase by reusing existing code and maintaining the integrity of the core system. This sometimes makes contributing in certain areas more challenging, and we understand how that can feel disheartening. That said, we always encourage new contributors and want everyone to feel that their efforts are valued, even if not all changes make it through.

Regarding attribution, I sincerely apologize if you felt overlooked. While Guy did express appreciation for your work in his comments, we also plan to publicly acknowledge your contributions on our Discord server with a thank-you announcement, highlighting the feature you’ve brought to life and the impact it has made.

We hope you’ll continue to be part of the ChartDB community, whether as a user or contributor. Your passion and ideas are truly valued, and we’d love to see more of your work in the future. That said, I want to be transparent that while contributions may not always get merged, they are appreciated and discussed thoroughly.

Thank you again for your contributions, and I’m here if you want to discuss further or share your thoughts.

Jonathan

johnnyfish commented:
Additionally, I’d like to share a note about our contribution process to help set expectations and avoid potential disappointments. While we recommend that contributors bring up features as issues first and wait for us to assign them after discussion - ensuring alignment on approach and considerations - we still welcome PRs and ideas that come in without a pre-discussed issue. This allows us to remain agile and foster creativity within the community.

However, please keep in mind that even if a PR is submitted without a prior issue or after assignment, there may still be cases where it does not meet the standards we strive to maintain for ChartDB, and **it may not be merged**. We hope this doesn’t deter you from contributing, as every idea and effort helps drive the project forward.

**Thank you again for your dedication and for being a valued part of our community.**

nikelborm commented:
@johnnyfish 

But why reject the whole PR entirely? We could easily collaborate and I could easily just correct the code to use new functions even though they were added just 2 hours before closing the PR. Why don't just update this PR as repo owners and add commits on top of those in this PR? Or why not ask for changes? Why slam-close the PR and create another one??

I finished working on the PR `Sun, 10 Nov 2024 01:19:10 +0300` even before there were anything even closely resembling cloning tables.

Then on Sun, `10 Nov 2024 16:30:15 +0200` there were added commit which still had no function for cloning tables, but there was functionality to cloneDiagram the whole diagram: https://github.com/chartdb/chartdb/commit/94a5d84fae819b0de6c1e471d1aad16dc8f39dd6#diff-da10bcc7d8dff796f50c9358a7268eb5abd2c50f460aecc4b2be1c691a966224R98

And only `Thu, 14 Nov 2024 15:11:21 +0200` it turned into cloneTable https://github.com/chartdb/chartdb/commit/9698821828a0475b13c7b2048606ec34e202d6c7#diff-8cbac28b5dad0659c5378293821f53f595d0884f9d1dbe7c8bad16b2a7f75fbbR49 right before (with a 2.3 hour difference: `Nov 14, 2024, 6:32 PM GMT+3`) shutting down this PR, which doesn't tell me anything other than about connection of those 2 events. It didn't happen just accidentally.

I just don't understand why do it that way?? If I did something wrong accidentally I'm happy to make an apology, but if not, I would be happy to know why was it done this way?

johnnyfish commented:
@nikelborm 

Thank you for your thoughtful and heartfelt response. **I want to start by sincerely acknowledging and appreciating the passion and dedication you’ve shown to improving ChartDB.** It’s clear that you put significant effort into your PR and have a genuine love for databases and building solutions in this space. Your motivation and expertise are highly valued.

I understand your frustration and confusion regarding the handling of your PR. **Closing it without collaborating further or providing a chance for you to make changes likely felt abrupt and discouraging. This was not our intention, and I’m truly sorry for the way it unfolded.**

The decision to close the PR was not meant to undermine your work, but rather, it stemmed from an effort to maintain code consistency and reusability within a rapidly evolving part of the codebase. **That said, I fully recognize that a more collaborative approach - such as suggesting changes, working together to adjust your code, or even merging your PR with follow-up modifications - would have been far more constructive. This is a learning moment for us as well, and I’m committed to improving our process to ensure contributors like yourself feel respected and supported.**

To address your questions: the timing of the changes and the introduction of the cloneTable function was genuinely coincidental and related to the fast pace of recent updates. I realize that this may have come across as dismissive of your efforts, and I regret that it felt this way. Your willingness to adapt and collaborate, even at short notice, is appreciated, and we should have leveraged that rather than closing the PR.

**Moving forward, I’d like to personally invite you to continue contributing and collaborating with us.** If you’re open to it, we’d love to revisit this feature together to see how we can incorporate your ideas and work in a meaningful way. Additionally, if you’re interested, I’d be happy to **set up a call or chat to discuss how we can make our collaboration smoother and more supportive in the future.** Your desire to help improve the project and potentially join the team is exciting, and we want to create a space where you feel valued and encouraged to contribute.

Thank you again for your openness and passion. Let’s keep the lines of communication open and work together to make ChartDB even better.

Jonathan & Guy.
